### DISCUSSION TITLE
FFmpeg NVENC: Add CUDA support for zero-copy 4:4:4 and RGB

### DISCUSSION THREAD

**Discussion by Xaymar** [Xaymar/obs-stream-effects#896] [Ideas & Feedback]
### Description
The NVENC encoders in FFmpeg support CUDA inputs, which also support 4:4:4 and RGBX formats. These two formats are not supported by the DirectX backend, or the OpenGL backend, so using CUDA benefits us here. We've already linked to CUDA through the use of the NVIDIA Face Tracking filter, and just have to extend the API a little bit.

### Checklist

- [ ] I am willing to pay someone to implement this feature.
- [x] This feature absolutely can't be done with already existing features.

**Comment by Xaymar** [comment]
Through further investigation, it has become clear that only two formats can be supported with zero-copy encoding: NV12, and RGBA - the other OBS video formats use additional textures which libOBS does not provide. Therefore in order to support this we would have to take in an RGBA texture and apply our own conversion to it, either as a CUDA kernel or as a Compute shader. The following conversions would need to be done:

- RGBA -> NV12 (OpenGL)
- RGBA -> YUV420 (both)
- RGBA -> YUV444 (both)
- RGBA -> XRGB (both)

At the current time, this is not yet planned. If OBS eventually supports P010 however it should be trivial to add support for encoding that using zero-copy methods.

**Reply by Xaymar** [reply]
The P010 format is available since OBS Studio 28.0, and allows for HDR capture. This is still a 4:2:0 format, so it will not allow for the unsupported formats to be used for encoding.

**Comment by massatt212** [comment]
i know this is not the place for a question, but does ffmpeg nvenc give a GTX 960 a boost in streaming quality or am i just wasting my time using it, cause i see turning mentioned alot.

**Comment by Fire-** [comment]
Is this the issue tracking the work required to support zero-copy lossless/perceptually-lossless encoding [as referenced in the blog](https://blog.xaymar.com/guides/high-quality-recordings/nvidia-nvenc/#:~:text=Effective%20lossless%20requires,support%20for%20it)?

**Comment by Xaymar** [comment]
This is the issue tracking support for NVENC 4:4:4 and RGB zero-copy capture.

**Comment by Fire-** [comment]
> This is the issue tracking support for NVENC 4:4:4 and RGB zero-copy capture.

Right, sorry, I guess I'm asking if this work is blocking the _ability_ to do zero-copy lossless recording - I'm sort of trying to figure out what the blockers for that functionality are to see if I'm able to implement any of it

**Comment by Xaymar** [comment]
As stated in the OP, this is the issue that tracks support for 4:4:4 and RGB zero-copy capture for NVENC using CUDA. CUDA is necessary due to [FFmpeg itself not supporting anything but NV12, 4:2:0 and P010 textures for the Direct3D11 backend](https://github.com/FFmpeg/FFmpeg/blob/master/libavutil/hwcontext_d3d11va.c#L82). With CUDA we can share [NV12, 4:2:0, A:4:2:0, 4:4:4, P010, P016, 4:4:4 P16, XRGB, XBGR and even Vulkan textures](https://github.com/FFmpeg/FFmpeg/blob/master/libavutil/hwcontext_cuda.c#L38). You can probably guess by now that without being able to share 4:4:4 or RGB textures, zero-copy capture for those will not work.

**Comment by Xaymar** [comment]
#### Rejected: Impossible
This may come as a shock to the ones following this issue, but it was inevitable once I started really investigating the issue. After a few minutes of delving today, I now know that zero-copy GPU encoding for anything but NV12 is impossible, as [libOBS does not include the code path required for it](https://github.com/obsproject/obs-studio/blob/master/libobs/obs-video-gpu-encode.c#L152-L197). Unfortunately that means that this feature is impossible until libOBS itself supports it.

Therefore, closing as impossible since upstream does not support it.

**Reply by Xaymar** [reply]
An incomplete Pull Request that addresses this shortcoming [is available](https://github.com/obsproject/obs-studio/pull/4974), however it appears to not have seen much activity.

**Comment by Fire-** [comment]
@Xaymar thanks for the update! Could you open an issue on the obs repo for support of that functionality so this might become implementable in the future? I don't have the contextual knowledge you possess in order to make it myself

**Comment by Xaymar** [comment]
@Fire- The upstream developers do not wish for me to contribute in any way, due to past differences over how they run the project.
